In [3]:
import pandas as pd
import numpy as np


# Работа с Pandas

# Загрузите ваш DataFrame в переменную df и выведете на экран первые пять строк 

In [4]:
df = pd.read_csv('https://raw.githubusercontent.com/W1llAnn/1tdata-homework/main/1.3googleplaystore.csv', delimiter=',')
df.head()

App        Category  Rating   
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1  \
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type  Price Content Rating   
0     159   19M      10,000+  Free    0.0       Everyone  \
1     967   14M     500,000+  Free    0.0       Everyone   
2   87510  8.7M   5,000,000+  Free    0.0       Everyone   
3  215644   25M  50,000,000+  Free    0.0           Teen   
4     967  2.8M     100,000+  Free    0.0       Everyone   

                      Genres      Last Updated         Current Ver   
0               Art & Design   January 7, 2018               1.0.0  \
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

# Отобразите количество строк и столбцов в DataFrame

In [5]:
df.shape

(10841, 13)

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  float64
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(2), object(11)
memory usage: 1.1+ MB


# PRO часть (Не обязательна к выполнению, но желательна :))


# Измените названия всех столбцов приведя их к нижнему регистру

In [7]:
df.columns = df.columns.str.lower()
df.columns = df.columns.str.replace(' ', '_')

**Проверяем:**

In [8]:
df.head()

app        category  rating   
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1  \
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  reviews  size     installs  type  price content_rating   
0     159   19M      10,000+  Free    0.0       Everyone  \
1     967   14M     500,000+  Free    0.0       Everyone   
2   87510  8.7M   5,000,000+  Free    0.0       Everyone   
3  215644   25M  50,000,000+  Free    0.0           Teen   
4     967  2.8M     100,000+  Free    0.0       Everyone   

                      genres      last_updated         current_ver   
0               Art & Design   January 7, 2018               1.0.0  \
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    android_ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

# Выведите основные статистики по категориальным столбцам. Напишите выводы о них 

**Сначала посмотрим общую информацию**

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   app             10841 non-null  object 
 1   category        10841 non-null  object 
 2   rating          9367 non-null   float64
 3   reviews         10841 non-null  object 
 4   size            10841 non-null  object 
 5   installs        10841 non-null  object 
 6   type            10840 non-null  object 
 7   price           10841 non-null  float64
 8   content_rating  10840 non-null  object 
 9   genres          10841 non-null  object 
 10  last_updated    10841 non-null  object 
 11  current_ver     10833 non-null  object 
 12  android_ver     10838 non-null  object 
dtypes: float64(2), object(11)
memory usage: 1.1+ MB


 - В датафтейме 13 столбцов
 - 2 из них имеют тип float64
 - 11 имеют тип object 
 - Всего строк - 10841 
 - Пропуски есть в колонках: Rating, Type, Content Rating, Current Ver и Android Ver
 - Объем занимаемый датафреймом 1.1+ MB

**Рассмотрим следующие категориальные столбцы: app, category, reviews, size, installs, type, content_rating, genres, last_updated, current_ver, android_ver**


In [10]:
df['app'].unique()

array(['Photo Editor & Candy Camera & Grid & ScrapBook',
       'Coloring book moana',
       'U Launcher Lite – FREE Live Cool Themes, Hide Apps', ...,
       'Parkinson Exercices FR', 'The SCP Foundation DB fr nn5n',
       'iHoroscope - 2018 Daily Horoscope & Astrology'], dtype=object)

In [11]:
df['category'].unique()

array(['ART_AND_DESIGN', 'AUTO_AND_VEHICLES', 'BEAUTY',
       'BOOKS_AND_REFERENCE', 'BUSINESS', 'COMICS', 'COMMUNICATION',
       'DATING', 'EDUCATION', 'ENTERTAINMENT', 'EVENTS', 'FINANCE',
       'FOOD_AND_DRINK', 'HEALTH_AND_FITNESS', 'HOUSE_AND_HOME',
       'LIBRARIES_AND_DEMO', 'LIFESTYLE', 'GAME', 'FAMILY', 'MEDICAL',
       'SOCIAL', 'SHOPPING', 'PHOTOGRAPHY', 'SPORTS', 'TRAVEL_AND_LOCAL',
       'TOOLS', 'PERSONALIZATION', 'PRODUCTIVITY', 'PARENTING', 'WEATHER',
       'VIDEO_PLAYERS', 'NEWS_AND_MAGAZINES', 'MAPS_AND_NAVIGATION',
       '1.9'], dtype=object)

In [12]:
df['reviews'].unique()

array(['159', '967', '87510', ..., '603', '1195', '398307'], dtype=object)

In [13]:
df['size'].unique()

array(['19M', '14M', '8.7M', '25M', '2.8M', '5.6M', '29M', '33M', '3.1M',
       '28M', '12M', '20M', '21M', '37M', '2.7M', '5.5M', '17M', '39M',
       '31M', '4.2M', '7.0M', '23M', '6.0M', '6.1M', '4.6M', '9.2M',
       '5.2M', '11M', '24M', 'Varies with device', '9.4M', '15M', '10M',
       '1.2M', '26M', '8.0M', '7.9M', '56M', '57M', '35M', '54M', '201k',
       '3.6M', '5.7M', '8.6M', '2.4M', '27M', '2.5M', '16M', '3.4M',
       '8.9M', '3.9M', '2.9M', '38M', '32M', '5.4M', '18M', '1.1M',
       '2.2M', '4.5M', '9.8M', '52M', '9.0M', '6.7M', '30M', '2.6M',
       '7.1M', '3.7M', '22M', '7.4M', '6.4M', '3.2M', '8.2M', '9.9M',
       '4.9M', '9.5M', '5.0M', '5.9M', '13M', '73M', '6.8M', '3.5M',
       '4.0M', '2.3M', '7.2M', '2.1M', '42M', '7.3M', '9.1M', '55M',
       '23k', '6.5M', '1.5M', '7.5M', '51M', '41M', '48M', '8.5M', '46M',
       '8.3M', '4.3M', '4.7M', '3.3M', '40M', '7.8M', '8.8M', '6.6M',
       '5.1M', '61M', '66M', '79k', '8.4M', '118k', '44M', '695k', '1.6M',
     

In [14]:
df['installs'].unique()

array(['10,000+', '500,000+', '5,000,000+', '50,000,000+', '100,000+',
       '50,000+', '1,000,000+', '10,000,000+', '5,000+', '100,000,000+',
       '1,000,000,000+', '1,000+', '500,000,000+', '50+', '100+', '500+',
       '10+', '1+', '5+', '0+', '0', 'Free'], dtype=object)

In [15]:
df['type'].unique()

array(['Free', 'Paid', nan, '0'], dtype=object)

In [16]:
df['content_rating'].unique()

array(['Everyone', 'Teen', 'Everyone 10+', 'Mature 17+',
       'Adults only 18+', 'Unrated', nan], dtype=object)

In [17]:
df['genres'].unique()

array(['Art & Design', 'Art & Design;Pretend Play',
       'Art & Design;Creativity', 'Art & Design;Action & Adventure',
       'Auto & Vehicles', 'Beauty', 'Books & Reference', 'Business',
       'Comics', 'Comics;Creativity', 'Communication', 'Dating',
       'Education;Education', 'Education', 'Education;Creativity',
       'Education;Music & Video', 'Education;Action & Adventure',
       'Education;Pretend Play', 'Education;Brain Games', 'Entertainment',
       'Entertainment;Music & Video', 'Entertainment;Brain Games',
       'Entertainment;Creativity', 'Events', 'Finance', 'Food & Drink',
       'Health & Fitness', 'House & Home', 'Libraries & Demo',
       'Lifestyle', 'Lifestyle;Pretend Play',
       'Adventure;Action & Adventure', 'Arcade', 'Casual', 'Card',
       'Casual;Pretend Play', 'Action', 'Strategy', 'Puzzle', 'Sports',
       'Music', 'Word', 'Racing', 'Casual;Creativity',
       'Casual;Action & Adventure', 'Simulation', 'Adventure', 'Board',
       'Trivia', 'Role 

In [18]:
df['last_updated'].unique()

array(['January 7, 2018', 'January 15, 2018', 'August 1, 2018', ...,
       'January 20, 2014', 'February 16, 2014', 'March 23, 2014'],
      dtype=object)

In [19]:
df['current_ver'].unique()

array(['1.0.0', '2.0.0', '1.2.4', ..., '1.0.612928', '0.3.4', '2.0.148.0'],
      dtype=object)

In [20]:
df['android_ver'].unique()

array(['4.0.3 and up', '4.2 and up', '4.4 and up', '2.3 and up',
       '3.0 and up', '4.1 and up', '4.0 and up', '2.3.3 and up',
       'Varies with device', '2.2 and up', '5.0 and up', '6.0 and up',
       '1.6 and up', '1.5 and up', '2.1 and up', '7.0 and up',
       '5.1 and up', '4.3 and up', '4.0.3 - 7.1.1', '2.0 and up',
       '3.2 and up', '4.4W and up', '7.1 and up', '7.0 - 7.1.1',
       '8.0 and up', '5.0 - 8.0', '3.1 and up', '2.0.1 and up',
       '4.1 - 7.1.1', nan, '5.0 - 6.0', '1.0 and up', '2.2 - 7.1.1',
       '5.0 - 7.1.1'], dtype=object)

**Посмотрим количество уникальных значений в наших колонках app, category, reviews, size, installs, type, content_rating, genres, last_updated, current_ver, android_ver**

In [21]:
print(f"количество уникальных значений в app - {df['app'].nunique()}")
print(f"количество уникальных значений в category - {df['category'].nunique()}")
print(f"количество уникальных значений в reviews - {df['reviews'].nunique()}")
print(f"количество уникальных значений в size - {df['size'].nunique()}")
print(f"количество уникальных значений в installs - {df['installs'].nunique()}")
print(f"количество уникальных значений в type - {df['type'].nunique()}")
print(f"количество уникальных значений в content_rating - {df['content_rating'].nunique()}")
print(f"количество уникальных значений в genres - {df['genres'].nunique()}")
print(f"количество уникальных значений в last_updated - {df['last_updated'].nunique()}")
print(f"количество уникальных значений в current_ver - {df['current_ver'].nunique()}")
print(f"количество уникальных значений в android_ver - {df['android_ver'].nunique()}")

количество уникальных значений в app - 9660
количество уникальных значений в category - 34
количество уникальных значений в reviews - 6002
количество уникальных значений в size - 462
количество уникальных значений в installs - 22
количество уникальных значений в type - 3
количество уникальных значений в content_rating - 6
количество уникальных значений в genres - 120
количество уникальных значений в last_updated - 1378
количество уникальных значений в current_ver - 2832
количество уникальных значений в android_ver - 33


# Выведите на экран часть датафрейма, который соответствует условию: `приложение из категории искусства и дизайна` И `с рейтингом больше 4.5`

In [22]:
df_filtered = df[(df['genres'] == 'Art & Design') & (df['rating'] > 4.5)]
df_filtered

app        category   
2     U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN  \
13                                Mandala Coloring Book  ART_AND_DESIGN   
16         Photo Designer - Write your name with shapes  ART_AND_DESIGN   
19                                         ibis Paint X  ART_AND_DESIGN   
22              Superheroes Wallpapers | 4K Backgrounds  ART_AND_DESIGN   
24                          HD Mickey Minnie Wallpapers  ART_AND_DESIGN   
25                           Harley Quinn wallpapers HD  ART_AND_DESIGN   
34                                      I Creative Idea  ART_AND_DESIGN   
38                       PIP Camera - PIP Collage Maker  ART_AND_DESIGN   
45    Canva: Poster, banner, card maker & graphic de...  ART_AND_DESIGN   
46    Install images with music to make video withou...  ART_AND_DESIGN   
3982                                  Cardi B Wallpaper  ART_AND_DESIGN   
4749  X Launcher: With OS11 Style Theme & Control Ce...  ART_AND_DESIGN   
4755  X Launcher Pro: PhoneX Theme, OS11 Control Center  ART_AND_DESIGN   
4759  X Launcher Pro - IOS Style Theme & Control Center  ART_AND_DESIGN   
4764     X Launcher Prime: With OS Style Theme & No Ads  ART_AND_DESIGN   
5254                            AJ Styles HD Wallpapers  ART_AND_DESIGN   
6930               Fantasy theme dark bw black building  ART_AND_DESIGN   
8888            Spring flowers theme couleurs d t space  ART_AND_DESIGN   

      rating reviews  size     installs  type  price content_rating   
2        4.7   87510  8.7M   5,000,000+  Free   0.00       Everyone  \
13       4.6    4326   21M     100,000+  Free   0.00       Everyone   
16       4.7    3632  5.5M     500,000+  Free   0.00       Everyone   
19       4.6  224399   31M  10,000,000+  Free   0.00       Everyone   
22       4.7    7699  4.2M     500,000+  Free   0.00   Everyone 10+   
24       4.7     118   23M      50,000+  Free   0.00       Everyone   
25       4.8     192  6.0M      10,000+  Free   0.00       Everyone   
34       4.7     353  4.2M      10,000+  Free   0.00           Teen   
38       4.7     158   11M      10,000+  Free   0.00       Everyone   
45       4.7  174531   24M  10,000,000+  Free   0.00       Everyone   
46       4.6    1070   26M     100,000+  Free   0.00       Everyone   
3982     4.8     253  3.7M      50,000+  Free   0.00       Everyone   
4749     4.7    5754  4.4M     100,000+  Free   0.00       Everyone   
4755     4.7     801  3.5M       5,000+  Paid   1.99       Everyone   
4759     4.8    1216  8.6M      10,000+  Paid   1.99       Everyone   
4764     4.7     149  3.5M       1,000+  Paid   1.99       Everyone   
5254     4.8      48   25M       5,000+  Free   0.00       Everyone   
6930     4.8      41  1.9M       5,000+  Free   0.00       Everyone   
8888     5.0       1  2.9M         100+  Free   0.00       Everyone   

            genres       last_updated current_ver   android_ver  
2     Art & Design     August 1, 2018       1.2.4  4.0.3 and up  
13    Art & Design      June 26, 2018       1.0.4    4.4 and up  
16    Art & Design      July 31, 2018         3.1    4.1 and up  
19    Art & Design      July 30, 2018       5.5.4    4.1 and up  
22    Art & Design      July 12, 2018     2.2.6.2  4.0.3 and up  
24    Art & Design       July 7, 2018       1.1.3    4.1 and up  
25    Art & Design     April 25, 2018         1.5    3.0 and up  
34    Art & Design     April 27, 2018         1.6    4.1 and up  
38    Art & Design  November 29, 2017         1.3  4.0.3 and up  
45    Art & Design      July 31, 2018       1.6.1    4.1 and up  
46    Art & Design  November 14, 2017         1.6    4.1 and up  
3982  Art & Design   November 1, 2017       1.0.0    4.0 and up  
4749  Art & Design      July 30, 2018       2.1.2    4.1 and up  
4755  Art & Design      July 30, 2018       2.1.2    4.1 and up  
4759  Art & Design      June 26, 2018       1.0.0    4.1 and up  
4764  Art & Design      July 30, 2018       1.2.2    4.1 and up  
5254  Ar

In [23]:
df_filtered.info()

<class 'pandas.core.frame.DataFrame'>
Index: 19 entries, 2 to 8888
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   app             19 non-null     object 
 1   category        19 non-null     object 
 2   rating          19 non-null     float64
 3   reviews         19 non-null     object 
 4   size            19 non-null     object 
 5   installs        19 non-null     object 
 6   type            19 non-null     object 
 7   price           19 non-null     float64
 8   content_rating  19 non-null     object 
 9   genres          19 non-null     object 
 10  last_updated    19 non-null     object 
 11  current_ver     19 non-null     object 
 12  android_ver     19 non-null     object 
dtypes: float64(2), object(11)
memory usage: 2.1+ KB


**Осталось 19 строк, которые подходят к условиям.**

# Посчитатайте какое количество приложений содержится в рейтенге, используя параметр разбиения (bins) равным 100. Выведите на экран топ-10 игр по количеству в рейтинге. Напишите краткий вывод.

In [24]:
print(f"Всего количество уникальных значений в app - {df['app'].nunique()}")

Всего количество уникальных значений в app - 9660


In [38]:
df_fltgame = df[df['category'] == 'GAME']
df_fltgame

app category  rating   reviews size   
1653                         ROBLOX     GAME     4.5   4447388  67M  \
1654                 Subway Surfers     GAME     4.5  27722264  76M   
1655               Candy Crush Saga     GAME     4.4  22426677  74M   
1656                      Solitaire     GAME     4.7    254258  23M   
1657                 Bubble Shooter     GAME     4.5    148897  46M   
...                             ...      ...     ...       ...  ...   
10791             Winter Wonderland     GAME     4.0      1287  38M   
10792        Soccer Clubs Logo Quiz     GAME     4.2     21661  16M   
10793                     Sid Story     GAME     4.4     28510  78M   
10803  Fatal Raid - No.1 Mobile FPS     GAME     4.3     56496  81M   
10804                  Poker Pro.Fr     GAME     4.2      5442  17M   

             installs  type  price content_rating   
1653     100,000,000+  Free    0.0   Everyone 10+  \
1654   1,000,000,000+  Free    0.0   Everyone 10+   
1655     500,000,000+  Free    0.0       Everyone   
1656      10,000,000+  Free    0.0       Everyone   
1657      10,000,000+  Free    0.0       Everyone   
...               ...   ...    ...            ...   
10791         50,000+  Free    0.0       Everyone   
10792      1,000,000+  Free    0.0       Everyone   
10793        500,000+  Free    0.0           Teen   
10803      1,000,000+  Free    0.0           Teen   
10804        100,000+  Free    0.0           Teen   

                             genres       last_updated   current_ver   
1653   Adventure;Action & Adventure      July 31, 2018  2.347.225742  \
1654                         Arcade      July 12, 2018        1.90.0   
1655                         Casual       July 5, 2018     1.129.0.2   
1656                           Card     August 1, 2018       2.137.0   
1657                         Casual      July 17, 2018        1.20.1   
...                             ...                ...           ...   
10791                          Word  December 18, 2013           1.0   
10792                        Trivia       May 24, 2018        1.3.81   
10793                          Card     August 1, 2018         2.6.6   
10803                        Action     August 7, 2018       1.5.447   
10804                          Card       May 22, 2018         4.1.3   

        android_ver  
1653     4.1 and up  
1654     4.1 and up  
1655     4.1 and up  
1656     4.1 and up  
1657   4.0.3 and up  
...             ...  
10791    2.2 and up  
10792    4.0 and up  
10793  4.0.3 and up  
10803    4.0 and up  
10804    2.3 and up  

[1144 rows x 13 columns]

In [99]:
df_fltgame['rating'].value_counts(bins=100)

rating
(4.48, 4.52]    174
(4.36, 4.4]     166
(4.28, 4.32]    151
(4.56, 4.6]     121
(4.16, 4.2]     115
               ... 
(3.04, 3.08]      0
(1.04, 1.08]      0
(1.84, 1.88]      0
(2.92, 2.96]      0
(3.0, 3.04]       0
Name: count, Length: 100, dtype: int64

**Прочитаем что получилось:**
**174 игры имеют рейтинг 4.48-4.52**
**166 игр имеют рейтинг 4.36-4.4**
**и так далее**

In [108]:
rsltg_df = df_fltgame.sort_values(by = 'rating', ascending = False)
rsltg_df[:10]

app category  rating reviews   
7698                                     CP Trivia     GAME     5.0       5  \
5917                                      Ra Ga Ba     GAME     5.0       2   
9056                   Santa's Monster Shootout DX     GAME     5.0       4   
5795                              Axe Champs! Wars     GAME     5.0       8   
6700                              Brick Breaker BR     GAME     5.0       7   
7466                                        211:CK     GAME     5.0       8   
5822   Flippy Axe : Flip The Knife & Axe Simulator     GAME     5.0       7   
9265                                      EC Mover     GAME     5.0       5   
7402                          Trovami se ci riesci     GAME     5.0      11   
10721                            Mad Dash Fo' Cash     GAME     5.0      14   

       size installs  type  price content_rating  genres        last_updated   
7698    12M     100+  Free   0.00       Everyone  Trivia      August 6, 2018  \
5917    20M       1+  Paid   1.49       Everyone  Arcade    February 8, 2017   
9056    33M      50+  Paid   1.99           Teen  Action     August 15, 2013   
5795    25M      50+  Free   0.00       Everyone  Arcade       June 26, 2018   
6700    19M       5+  Free   0.00       Everyone  Arcade       July 23, 2018   
7466    38M      10+  Paid   0.99           Teen  Arcade      April 11, 2018   
5822    15M     100+  Free   0.00       Everyone  Arcade  September 27, 2017   
9265   4.6M      10+  Free   0.00       Everyone  Racing      August 1, 2018   
7402   6.1M      10+  Free   0.00       Everyone  Arcade      March 11, 2017   
10721   16M     100+  Free   0.00       Everyone  Arcade       June 19, 2017   

      current_ver   android_ver  
7698         0.99    5.0 and up  
5917        1.0.4    2.3 and up  
9056         1.05    2.2 and up  
5795          1.1    4.1 and up  
6700          1.0    4.1 and up  
7466          1.3    4.1 and up  
5822          1.1    4.2 and up  
9265         1.11  4.0.3 and up  
7402          0.1    2.3 and up  
10721        2.5a    4.1 and up

**Можно все предыдущее объединить в одно**

In [103]:
df_fltgame2 = df[df['category'] == 'GAME'].sort_values(by = 'rating', ascending = False)
df_fltgame2[:10]

app category  rating reviews   
7698                                     CP Trivia     GAME     5.0       5  \
5917                                      Ra Ga Ba     GAME     5.0       2   
9056                   Santa's Monster Shootout DX     GAME     5.0       4   
5795                              Axe Champs! Wars     GAME     5.0       8   
6700                              Brick Breaker BR     GAME     5.0       7   
7466                                        211:CK     GAME     5.0       8   
5822   Flippy Axe : Flip The Knife & Axe Simulator     GAME     5.0       7   
9265                                      EC Mover     GAME     5.0       5   
7402                          Trovami se ci riesci     GAME     5.0      11   
10721                            Mad Dash Fo' Cash     GAME     5.0      14   

       size installs  type  price content_rating  genres        last_updated   
7698    12M     100+  Free   0.00       Everyone  Trivia      August 6, 2018  \
5917    20M       1+  Paid   1.49       Everyone  Arcade    February 8, 2017   
9056    33M      50+  Paid   1.99           Teen  Action     August 15, 2013   
5795    25M      50+  Free   0.00       Everyone  Arcade       June 26, 2018   
6700    19M       5+  Free   0.00       Everyone  Arcade       July 23, 2018   
7466    38M      10+  Paid   0.99           Teen  Arcade      April 11, 2018   
5822    15M     100+  Free   0.00       Everyone  Arcade  September 27, 2017   
9265   4.6M      10+  Free   0.00       Everyone  Racing      August 1, 2018   
7402   6.1M      10+  Free   0.00       Everyone  Arcade      March 11, 2017   
10721   16M     100+  Free   0.00       Everyone  Arcade       June 19, 2017   

      current_ver   android_ver  
7698         0.99    5.0 and up  
5917        1.0.4    2.3 and up  
9056         1.05    2.2 and up  
5795          1.1    4.1 and up  
6700          1.0    4.1 and up  
7466          1.3    4.1 and up  
5822          1.1    4.2 and up  
9265         1.11  4.0.3 and up  
7402          0.1    2.3 and up  
10721        2.5a    4.1 and up

# Создайте функцию , которая будет применена к столбцу рейтинг методом `apply()`.
Функция должна принимать строку с рейтингом и возвращать:
 - High rating, если рейтинг больше или равен 4.5
 - Middle rating, если рейтинг между 3.8 и 4.5
 - Low rating, если рейтинг меньше 3.8 
Результат работы функции запишите в новый столбец с названием `categorical_rating`

In [62]:
def make_rating(row):
    if row >= 4.5:
        return 'High rating'
    elif row < 4.5 and row >= 3.8:
        return 'Middle rating'
    elif row < 3.8:
        return 'Low rating'
    else:
        return 'unknown'

In [69]:
df_rat = df.copy()
df_rat['rating_class'] = df_rat['rating'].apply(make_rating)


df_rat

app             category   
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN  \
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       rating reviews                size     installs  type  price   
0         4.1     159                 19M      10,000+  Free    0.0  \
1         3.9     967                 14M     500,000+  Free    0.0   
2         4.7   87510                8.7M   5,000,000+  Free    0.0   
3         4.5  215644                 25M  50,000,000+  Free    0.0   
4         4.3     967                2.8M     100,000+  Free    0.0   
...       ...     ...                 ...          ...   ...    ...   
10836     4.5      38                 53M       5,000+  Free    0.0   
10837     5.0       4                3.6M         100+  Free    0.0   
10838     NaN       3                9.5M       1,000+  Free    0.0   
10839     4.5     114  Varies with device       1,000+  Free    0.0   
10840     4.5  398307                 19M  10,000,000+  Free    0.0   

      content_rating                     genres      last_updated   
0           Everyone               Art & Design   January 7, 2018  \
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              current_ver         android_ver   rating_class  
0                   1.0.0        4.0.3 and up  Middle rating  
1                   2.0.0        4.0.3 and up  Middle rating  
2                   1.2.4        4.0.3 and up    High rating  
3      Varies with device          4.2 and up    High rating  
4                     1.1          4.4 and up  Middle rating  
...                   ...                 ...            ...  
10836                1.48          4.1 and up    High rating  
10837                 1.0          4.1 and up    High rating  
10838                 1.0          2.2 and up        unknown  
10839  Varies with device  Varies with device    High rating  
10840  Varies with device  Varies with device    High rating  

[10841 rows x 14 columns]

In [86]:
print(f"Middle rating {len(df_rat[df_rat['rating_class'] == 'Middle rating'])}")
print(f"High rating {len(df_rat[df_rat['rating_class'] == 'High rating'])}")
print(f"Low rating {len(df_rat[df_rat['rating_class'] == 'Low rating'])}")
print(f"unknown {len(df_rat[df_rat['rating_class'] == 'unknown'])}")


Middle rating 5102
High rating 2956
Low rating 1309
unknown 1474


# Выведите кол-во уникальных значений в столбцах отсортированных по убыванию на экран:

In [39]:
df.describe(include='object')

app category reviews                size    installs   type   
count    10841    10841   10841               10841       10841  10840  \
unique    9660       34    6002                 462          22      3   
top     ROBLOX   FAMILY       0  Varies with device  1,000,000+   Free   
freq         9     1972     596                1695        1579  10039   

       content_rating genres    last_updated         current_ver android_ver  
count           10840  10841           10841               10833       10838  
unique              6    120            1378                2832          33  
top          Everyone  Tools  August 3, 2018  Varies with device  4.1 and up  
freq             8714    842             326                1459        2451

**count - Первая строка - посчитаны значения по столбцам с типом object**
**unique - Сколько уникальных значений по столбцам с типом object**
**top - Самое большое значение по столбцам с типом object**
**freq - Частота - это частота наиболее распространенного значения**

**Так как не указаны конкретные столбцы, сделаем по всем по отдельности**

***app***

In [96]:
df.app.value_counts()

app
ROBLOX                                                9
CBS Sports App - Scores, News, Stats & Watch Live     8
ESPN                                                  7
Duolingo: Learn Languages Free                        7
Candy Crush Saga                                      7
                                                     ..
Meet U - Get Friends for Snapchat, Kik & Instagram    1
U-Report                                              1
U of I Community Credit Union                         1
Waiting For U Launcher Theme                          1
iHoroscope - 2018 Daily Horoscope & Astrology         1
Name: count, Length: 9660, dtype: int64

***category***

In [41]:
df.category.value_counts()

category
FAMILY                 1972
GAME                   1144
TOOLS                   843
MEDICAL                 463
BUSINESS                460
PRODUCTIVITY            424
PERSONALIZATION         392
COMMUNICATION           387
SPORTS                  384
LIFESTYLE               382
FINANCE                 366
HEALTH_AND_FITNESS      341
PHOTOGRAPHY             335
SOCIAL                  295
NEWS_AND_MAGAZINES      283
SHOPPING                260
TRAVEL_AND_LOCAL        258
DATING                  234
BOOKS_AND_REFERENCE     231
VIDEO_PLAYERS           175
EDUCATION               156
ENTERTAINMENT           149
MAPS_AND_NAVIGATION     137
FOOD_AND_DRINK          127
HOUSE_AND_HOME           88
LIBRARIES_AND_DEMO       85
AUTO_AND_VEHICLES        85
WEATHER                  82
ART_AND_DESIGN           65
EVENTS                   64
PARENTING                60
COMICS                   60
BEAUTY                   53
1.9                       1
Name: count, dtype: int64

***rating***

In [42]:
df.rating.value_counts()

rating
4.4     1109
4.3     1076
4.5     1038
4.2      952
4.6      823
4.1      708
4.0      568
4.7      499
3.9      386
3.8      303
5.0      274
3.7      239
4.8      234
3.6      174
3.5      163
3.4      128
3.3      102
4.9       87
3.0       83
3.1       69
3.2       64
2.9       45
2.8       42
2.7       25
2.6       25
2.5       21
2.3       20
2.4       19
1.0       16
2.2       14
1.9       13
2.0       12
1.7        8
1.8        8
2.1        8
1.6        4
1.4        3
1.5        3
1.2        1
19.0       1
Name: count, dtype: int64

***reviews***

In [43]:
df.reviews.value_counts()

reviews
0         596
1         272
2         214
3         175
4         137
         ... 
342912      1
4272        1
5517        1
4057        1
398307      1
Name: count, Length: 6002, dtype: int64

***size***

In [47]:
df['size'].value_counts()

size
Varies with device    1695
11M                    198
12M                    196
14M                    194
13M                    191
                      ... 
429k                     1
200k                     1
460k                     1
728k                     1
619k                     1
Name: count, Length: 462, dtype: int64

***installs***

In [45]:
df.installs.value_counts()

installs
1,000,000+        1579
10,000,000+       1252
100,000+          1169
10,000+           1054
1,000+             907
5,000,000+         752
100+               719
500,000+           539
50,000+            479
5,000+             477
100,000,000+       409
10+                386
500+               330
50,000,000+        289
50+                205
5+                  82
500,000,000+        72
1+                  67
1,000,000,000+      58
0+                  14
0                    1
Free                 1
Name: count, dtype: int64

***type***

In [48]:
df.type.value_counts()

type
Free    10039
Paid      800
0           1
Name: count, dtype: int64

***price***

In [49]:
df.price.value_counts()

price
0.00     10041
0.99       148
2.99       129
1.99        73
4.99        72
         ...  
19.90        1
1.75         1
14.00        1
4.85         1
1.04         1
Name: count, Length: 92, dtype: int64

***content_rating***

In [50]:
df.content_rating.value_counts()



content_rating
Everyone           8714
Teen               1208
Mature 17+          499
Everyone 10+        414
Adults only 18+       3
Unrated               2
Name: count, dtype: int64

***genres***

In [51]:
df.genres.value_counts()

genres
Tools                     842
Entertainment             623
Education                 549
Medical                   463
Business                  460
                         ... 
Arcade;Pretend Play         1
Card;Brain Games            1
Lifestyle;Pretend Play      1
Comics;Creativity           1
Strategy;Creativity         1
Name: count, Length: 120, dtype: int64

***last_updated***

In [52]:
df.last_updated.value_counts()

last_updated
August 3, 2018        326
August 2, 2018        304
July 31, 2018         294
August 1, 2018        285
July 30, 2018         211
                     ... 
March 20, 2014          1
April 7, 2015           1
September 22, 2014      1
October 3, 2015         1
March 23, 2014          1
Name: count, Length: 1378, dtype: int64

***current_ver***

In [53]:
df.current_ver.value_counts()

current_ver
Varies with device    1459
1.0                    809
1.1                    264
1.2                    178
2.0                    151
                      ... 
1.0.17.3905              1
15.1.2                   1
4.94.19                  1
1.1.11.11                1
2.0.148.0                1
Name: count, Length: 2832, dtype: int64

***android_ver***

In [56]:
df.android_ver.value_counts()

android_ver
4.1 and up            2451
4.0.3 and up          1501
4.0 and up            1375
Varies with device    1362
4.4 and up             980
2.3 and up             652
5.0 and up             601
4.2 and up             394
2.3.3 and up           281
2.2 and up             244
4.3 and up             243
3.0 and up             241
2.1 and up             134
1.6 and up             116
6.0 and up              60
7.0 and up              42
3.2 and up              36
2.0 and up              32
5.1 and up              24
1.5 and up              20
4.4W and up             12
3.1 and up              10
2.0.1 and up             7
8.0 and up               6
7.1 and up               3
4.0.3 - 7.1.1            2
5.0 - 8.0                2
1.0 and up               2
7.0 - 7.1.1              1
4.1 - 7.1.1              1
5.0 - 6.0                1
2.2 - 7.1.1              1
5.0 - 7.1.1              1
Name: count, dtype: int64

**Посмотрим самое большое количество уникальных значений среди всех столбцов и строк**

In [88]:
counts_app = df['app'].value_counts()
counts_app = counts_app.to_frame()
counts_app['type'] = 'app'

counts_category = df['category'].value_counts()
counts_category = counts_category.to_frame()
counts_category['type'] = 'category'

counts_rating = df['rating'].value_counts()
counts_rating = counts_rating.to_frame()
counts_rating['type'] = 'rating'

counts_reviews = df['reviews'].value_counts()
counts_reviews = counts_reviews.to_frame()
counts_reviews['type'] = 'reviews'

counts_size = df['size'].value_counts()
counts_size = counts_size.to_frame()
counts_size['type'] = 'size'

counts_installs = df['installs'].value_counts()
counts_installs = counts_installs.to_frame()
counts_installs['type'] = 'installs'

counts_type = df['type'].value_counts()
counts_type = counts_type.to_frame()
counts_type['type'] = 'type'

counts_price = df['price'].value_counts()
counts_price = counts_price.to_frame()
counts_price['type'] = 'price'

counts_content_rating = df['content_rating'].value_counts()
counts_content_rating = counts_content_rating.to_frame()
counts_content_rating['type'] = 'content_rating'

counts_genres = df['genres'].value_counts()
counts_genres = counts_genres.to_frame()
counts_genres['type'] = 'genres'

counts_last_updated = df['last_updated'].value_counts()
counts_last_updated = counts_last_updated.to_frame()
counts_last_updated['type'] = 'last_updated'

counts_current_ver = df['current_ver'].value_counts()
counts_current_ver = counts_current_ver.to_frame()
counts_current_ver['type'] = 'current_ver'

counts_android_ver = df['android_ver'].value_counts()
counts_android_ver = counts_android_ver.to_frame()
counts_android_ver['type'] = 'android_ver'


chek = pd.concat([counts_app, counts_category, counts_rating, counts_reviews, counts_size, counts_installs, counts_type, counts_price, counts_content_rating, counts_genres, counts_last_updated, counts_current_ver, counts_android_ver]) 

chek

count         type
ROBLOX                                                 9          app
CBS Sports App - Scores, News, Stats & Watch Live      8          app
ESPN                                                   7          app
Duolingo: Learn Languages Free                         7          app
Candy Crush Saga                                       7          app
...                                                  ...          ...
7.0 - 7.1.1                                            1  android_ver
4.1 - 7.1.1                                            1  android_ver
5.0 - 6.0                                              1  android_ver
2.2 - 7.1.1                                            1  android_ver
5.0 - 7.1.1                                            1  android_ver

[20684 rows x 2 columns]

**Теперь отсортируем и посмотрим первые 10**

In [92]:
rslt_df = chek.sort_values(by = 'count', ascending = False)
rslt_df[:10]

count            type
0.0                 10041           price
Free                10039            type
Everyone             8714  content_rating
4.1 and up           2451     android_ver
FAMILY               1972        category
Varies with device   1695            size
1,000,000+           1579        installs
4.0.3 and up         1501     android_ver
Varies with device   1459     current_ver
4.0 and up           1375     android_ver

# И тут я поняла, что искала не просто уникальные значения, а уникальные и часто используемые в датафрейме... 

**Из таблицы сверху в принципе все понятно в столбце price значение 0.0 использовалось 10041 раз и так далее**

# __Super-Hard задание__

# Решите прошлую задачу, но вместо функции `def` используйте `лямбда-функцию` внутри `apply()`

In [90]:
df_rat2 = df.copy()
df_rat2['rating_class'] = df_rat2['rating'].apply(lambda x: 'High rating' if x >= 4.5 else ('Middle rating' if x < 4.5 and x >= 3.8 else ('Low rating' if x < 3.8 else 'unknown' )))

df_rat2

app             category   
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN  \
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       rating reviews                size     installs  type  price   
0         4.1     159                 19M      10,000+  Free    0.0  \
1         3.9     967                 14M     500,000+  Free    0.0   
2         4.7   87510                8.7M   5,000,000+  Free    0.0   
3         4.5  215644                 25M  50,000,000+  Free    0.0   
4         4.3     967                2.8M     100,000+  Free    0.0   
...       ...     ...                 ...          ...   ...    ...   
10836     4.5      38                 53M       5,000+  Free    0.0   
10837     5.0       4                3.6M         100+  Free    0.0   
10838     NaN       3                9.5M       1,000+  Free    0.0   
10839     4.5     114  Varies with device       1,000+  Free    0.0   
10840     4.5  398307                 19M  10,000,000+  Free    0.0   

      content_rating                     genres      last_updated   
0           Everyone               Art & Design   January 7, 2018  \
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              current_ver         android_ver   rating_class  
0                   1.0.0        4.0.3 and up  Middle rating  
1                   2.0.0        4.0.3 and up  Middle rating  
2                   1.2.4        4.0.3 and up    High rating  
3      Varies with device          4.2 and up    High rating  
4                     1.1          4.4 and up  Middle rating  
...                   ...                 ...            ...  
10836                1.48          4.1 and up    High rating  
10837                 1.0          4.1 and up    High rating  
10838                 1.0          2.2 and up        unknown  
10839  Varies with device  Varies with device    High rating  
10840  Varies with device  Varies with device    High rating  

[10841 rows x 14 columns]

In [80]:
df_rat2['rating_class'].unique()

array(['Middle rating', 'High rating', 'Low rating', 'unknown'],
      dtype=object)

In [85]:
print(f"Middle rating {len(df_rat2[df_rat2['rating_class'] == 'Middle rating'])}")
print(f"High rating {len(df_rat2[df_rat2['rating_class'] == 'High rating'])}")
print(f"Low rating {len(df_rat2[df_rat2['rating_class'] == 'Low rating'])}")
print(f"unknown {len(df_rat2[df_rat2['rating_class'] == 'unknown'])}")


Middle rating 5102
High rating 2956
Low rating 1309
unknown 1474


**Сравним с выводом нашей фунции:**

Middle rating 5102
High rating 2956
Low rating 1309
unknown 1474


**Сходится, значит lambda отработала как и функция**

# Изучите функцию `transform` в документации Pandas. Примените её к датафрейму, который будет заранее сгруппирован по столбцу `category`. С помощью фукнции `transform` найдите средний `rating` по каждой категории и запишите в столбец с названием `mean_cat_rating`. Кол-во строк в получившемся новом столбец должно быть таким же, как и в исходном датафрейме

In [117]:
df_transform = df.copy()
df_transform["mean_cat_rating"] = df_transform.groupby('category')["rating"].transform('mean')
df_transform

app             category   
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN  \
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       rating reviews                size     installs  type  price   
0         4.1     159                 19M      10,000+  Free    0.0  \
1         3.9     967                 14M     500,000+  Free    0.0   
2         4.7   87510                8.7M   5,000,000+  Free    0.0   
3         4.5  215644                 25M  50,000,000+  Free    0.0   
4         4.3     967                2.8M     100,000+  Free    0.0   
...       ...     ...                 ...          ...   ...    ...   
10836     4.5      38                 53M       5,000+  Free    0.0   
10837     5.0       4                3.6M         100+  Free    0.0   
10838     NaN       3                9.5M       1,000+  Free    0.0   
10839     4.5     114  Varies with device       1,000+  Free    0.0   
10840     4.5  398307                 19M  10,000,000+  Free    0.0   

      content_rating                     genres      last_updated   
0           Everyone               Art & Design   January 7, 2018  \
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              current_ver         android_ver  mean_cat_rating  
0                   1.0.0        4.0.3 and up         4.358065  
1                   2.0.0        4.0.3 and up         4.358065  
2                   1.2.4        4.0.3 and up         4.358065  
3      Varies with device          4.2 and up         4.358065  
4                     1.1          4.4 and up         4.358065  
...                   ...                 ...              ...  
10836                1.48          4.1 and up         4.192272  
10837                 1.0          4.1 and up         4.192272  
10838                 1.0          2.2 and up         4.189143  
10839  Varies with device  Varies with device         4.346067  
10840  Varies with device  Varies with device         4.094904  

[10841 rows x 14 columns]

**Столбец добавлен, количество строк не изменилось**

In [118]:
df_transform['category'].value_counts()

category
FAMILY                 1972
GAME                   1144
TOOLS                   843
MEDICAL                 463
BUSINESS                460
PRODUCTIVITY            424
PERSONALIZATION         392
COMMUNICATION           387
SPORTS                  384
LIFESTYLE               382
FINANCE                 366
HEALTH_AND_FITNESS      341
PHOTOGRAPHY             335
SOCIAL                  295
NEWS_AND_MAGAZINES      283
SHOPPING                260
TRAVEL_AND_LOCAL        258
DATING                  234
BOOKS_AND_REFERENCE     231
VIDEO_PLAYERS           175
EDUCATION               156
ENTERTAINMENT           149
MAPS_AND_NAVIGATION     137
FOOD_AND_DRINK          127
HOUSE_AND_HOME           88
LIBRARIES_AND_DEMO       85
AUTO_AND_VEHICLES        85
WEATHER                  82
ART_AND_DESIGN           65
EVENTS                   64
PARENTING                60
COMICS                   60
BEAUTY                   53
1.9                       1
Name: count, dtype: int64

In [119]:
df_transform['mean_cat_rating'].value_counts()

mean_cat_rating
4.192272     1972
4.286326     1144
4.047411      843
4.189143      463
4.121452      460
4.211396      424
4.335987      392
4.158537      387
4.223511      384
4.094904      382
4.131889      366
4.277104      341
4.192114      335
4.255598      295
4.132189      283
4.259664      260
4.109292      258
3.970769      234
4.346067      231
4.063750      175
4.389032      156
4.126174      149
4.051613      137
4.166972      127
4.197368       88
4.178462       85
4.190411       85
4.244000       82
4.358065       65
4.435556       64
4.300000       60
4.155172       60
4.278571       53
19.000000       1
Name: count, dtype: int64

**Эти данные тоже сходятся( на взгляд), но так и должно быть, так как groupby мы делали по category**

# Создайте функцию , которая будет применена к столбцу `last_updated` и возвращать из него число месяца, если число месяца вернуть из строки невозможно, то верните вместо этого стркоу с названием `miss_date`.\
# Запишите полученный столбец в датафрейм , название столбцу дайте `day_of_update`.

# Если вы столкнётесь с некоторой ошибкой , которую вам выдаст Pandas, то изучите конструкцию Python `try` \ `except`.\
# Вот псевдокод , который поможет лучше разобраться с этой конструкцией:

**Посмотрим часть уникальных значений**

In [120]:
list_all = df['last_updated'].unique()

array(['January 7, 2018', 'January 15, 2018', 'August 1, 2018', ...,
       'January 20, 2014', 'February 16, 2014', 'March 23, 2014'],
      dtype=object)

Как мы видим, формат записи данных состоит: Месяц День, Год. Нам нужен месяц

In [125]:
len(df['last_updated'].value_counts())

1378

In [10]:
month = ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December']

In [193]:
def l_updated(arrg):
    try:
        for data in range(1):
            str_upd = str(arrg[data]).split()

            if month.count(str_upd[0]):
                arrg
            else:
                pass
        return str(arrg).split()[0]

    except:
        return 'miss_date'


In [190]:
arr = df['last_updated'].unique()
l_updated(arr.flat[:5])

"['January"

функция работает на списке уникальных сначений

In [194]:
df['day_of_update'] = df['last_updated'].apply(l_updated)


In [195]:
df.head()

app        category  rating   
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1  \
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  reviews  size     installs  type  price content_rating   
0     159   19M      10,000+  Free    0.0       Everyone  \
1     967   14M     500,000+  Free    0.0       Everyone   
2   87510  8.7M   5,000,000+  Free    0.0       Everyone   
3  215644   25M  50,000,000+  Free    0.0           Teen   
4     967  2.8M     100,000+  Free    0.0       Everyone   

                      genres      last_updated         current_ver   
0               Art & Design   January 7, 2018               1.0.0  \
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    android_ver day_of_update  
0  4.0.3 and up       January  
1  4.0.3 and up       January  
2  4.0.3 and up        August  
3    4.2 and up          June  
4    4.4 and up          June